osf/management/commands/update_institution_sso_email_domain.py

```diff
@@ -0,0 +1,146 @@
+
+
+class UpdateResult(IntEnum):
+    """Defines 4 states of the email update outcome.
+    """
+    SUCCEEDED = 0  # The email has successfully been added
+    SKIPPED = 1  # The email was added (to the user) before the script is run, or the eligible email is not found
+    FAILED = 2  # The email failed to be added since it belongs to another account
+    ERRORED = 3  # The email failed to be added due to unexpected exceptions
+
+
+class Command(BaseCommand):
+    """Update the emails of users from the affiliated institution (when eligible). Requires three arguments:
+    `institution_id`, `src_domain` and `dst_domain`. The domains do not go through email validation or spam check
+    since the script can only be run by trusted and privileged admins.
+    """
+
+    def add_arguments(self, parser):
+        super(Command, self).add_arguments(parser)
+        parser.add_argument(
+            'institution_id',
+            type=str,
+            help='the institution whose affiliated users\' eligible email is to be updated'
+        )
+        parser.add_argument(
+            'src_domain',
+            type=str,
+            help='the email domain that are eligible for update'
+        )
+        parser.add_argument(
+            'dst_domain',
+            type=str,
+            help='the email domain that is to be added'
+        )
+        parser.add_argument(
+            '--dry',
+            action='store_true',
+            dest='dry_run',
+            help='If true, iterate through eligible users and emails but don\'t add the email'
+        )
+
+    def handle(self, *args, **options):
+
+        institution_id = options.get('institution_id', '').lower()
+        src_domain = options.get('src_domain', '').lower()
+        dst_domain = options.get('dst_domain', '').lower()
+        dry_run = options.get('dry_run', True)
+
+        if dry_run:
+            logger.warning('This is a dry-run pass.')
+
+        # Verify the institution
+        institution = Institution.load(institution_id)
+        if not institution:
+            message = 'Error: invalid institution ID [{}]'.format(institution_id)
+            logger.error(message)
+            sentry.log_message(message)
+            return
+
+        # Find all users affiliated with the given institution
+        affiliated_users = OSFUser.objects.filter(affiliated_institutions___id=institution_id)
+
+        # Update email domain for each user
+        update_result = {
+            UpdateResult.SUCCEEDED.name: {},
+            UpdateResult.SKIPPED.name: {},
+            UpdateResult.FAILED.name: {},
+            UpdateResult.ERRORED.name: {},
+        }
+        for user in affiliated_users:
+            user_result = update_email_domain(user, src_domain, dst_domain, dry_run=dry_run)
+            if user_result.get(UpdateResult.SUCCEEDED.name):
+                update_result.get(UpdateResult.SUCCEEDED.name)[user._id] = user_result.get(UpdateResult.SUCCEEDED.name)
```

Comment:
We iterate the users to add the email but the output would have been very difficult to read if it is keyed by the user on the top level. Thus, covert it to be keyed by the update status first and then by the user when generating the output.